File: typescript/src/session-recovery.ts
import { SYNTHETIC_CONTENT, type SessionState, type SyntheticReason, type ToolCallState } from "./session-reducer.js";

export { SYNTHETIC_CONTENT };
export type CurrentTool = {
    name: string;
    definitionDigest: string;
    replay: "safe" | "never";
    replayKey: string;
};
export type CurrentConfiguration = {
    configurationDigest: string;
    environmentIdentity: string;
    tools: CurrentTool[];
};
export type SyntheticResult = {
    assistantEntryId: string;
    toolIndex: number;
    toolCallId: string;
    toolName: string;
    toolStartedId?: string;
    resultEntryId?: string;
    reason: SyntheticReason;
    content: string;
};
export type RecoveryPlan =
    | {
          type: "finish";
          outcome: "completed" | "aborted" | "failed";
          completion?: "normal" | "truncated";
          finalEntryId?: string;
          error?: { code: string; message: string };
      }
    | {
          type: "startStep";
          stepKind: "assistant" | "compaction";
          attempt: 1 | 2;
          stepId?: string;
          contextThroughEntryId: string;
      }
    | {
          type: "startTool";
          mode: "start" | "replay";
          assistantEntryId: string;
          toolIndex: number;
          toolStartedId?: string;
          toolName: string;
          arguments: Record<string, unknown>;
      }
    | { type: "appendSynthetic"; results: SyntheticResult[] }
    | { type: "closeAttempt"; error: { code: "aborted"; message: "Operation aborted" } }
    | {
          type: "blocked";
          reason: "configuration_changed" | "environment_changed" | "replay_declaration_changed" | "attempts_exhausted";
      };

function toolDeclaration(current: CurrentConfiguration, name: string) {
    return current.tools.find((tool) => tool.name === name);
}

function startedSynthetic(tool: ToolCallState, reason: "interrupted"): SyntheticResult {
    return {
        assistantEntryId: tool.assistantEntryId,
        toolIndex: tool.toolIndex,
        toolCallId: tool.toolCallId,
        toolName: tool.toolName,
        toolStartedId: tool.toolStartedId,
        resultEntryId: tool.resultEntryId,
        reason,
        content: SYNTHETIC_CONTENT[reason],
    };
}

function assistantCalls(state: SessionState) {
    if (state.operation.kind !== "run" || !state.operation.step?.settledEntryId) return undefined;
    // The current run step is always settled by the transcript's most recent assistant-role
    // message: a new assistant turn can only be committed once every earlier assistant message's
    // tool calls are fully resolved. Filtering this scan down to "the last assistant message that
    // ever had tool calls" is wrong once the current turn's own response is a plain stop (no tool
    // calls) — it then falls back to an older, already-resolved turn and reports its calls as
    // still pending, which plans a stepAttempt against a now-stale contextThroughEntryId.
    const assistantIndex = state.transcript.findLastIndex((item) => item.role === "assistant");
    const message = state.transcript[assistantIndex];
    if (!message || message.role !== "assistant" || !message.tool_calls?.length) return undefined;
    const completedResults = state.transcript.slice(assistantIndex + 1).filter((item) => item.role === "tool").length;
    return {
        assistantEntryId: state.operation.step.settledEntryId,
        calls: message.tool_calls ?? [],
        completedResults,
    };
}

function preExecutionSynthetic(
    assistantEntryId: string,
    toolIndex: number,
    toolCallId: string,
    toolName: string,
    reason: Exclude<SyntheticReason, "interrupted">,
): SyntheticResult {
    return {
        assistantEntryId,
        toolIndex,
        toolCallId,
        toolName,
        reason,
        content: SYNTHETIC_CONTENT[reason],
    };
}

export function planRecovery(state: SessionState, current: CurrentConfiguration): RecoveryPlan {
    const operation = state.operation;
    if (operation.kind === "idle") return { type: "finish", outcome: "completed", completion: "normal" };

    const assistant = assistantCalls(state);
    const pending = operation.kind === "run" ? operation.toolCalls.filter((tool) => tool.status === "pending") : [];
    if (operation.abortRequested) {
        if (operation.step?.status === "attempting")
            return { type: "closeAttempt", error: { code: "aborted", message: "Operation aborted" } };
        if (operation.kind === "run" && assistant) {
            const started = new Map(
                operation.toolCalls.map((tool) => [`${tool.assistantEntryId}:${tool.toolIndex}`, tool]),
            );
            const results = assistant.calls.flatMap((call, toolIndex) => {
                if (toolIndex < assistant.completedResults) return [];
                const tool = started.get(`${assistant.assistantEntryId}:${toolIndex}`);
                if (tool?.status === "pending") return [startedSynthetic(tool, "interrupted")];
                if (tool) return [];
                return [
                    preExecutionSynthetic(
                        assistant.assistantEntryId,
                        toolIndex,
                        call.id,
                        call.function.name,
                        "aborted",
                    ),
                ];
            });
            if (results.length) return { type: "appendSynthetic", results };
        }
        return { type: "finish", outcome: "aborted" };
    }

    if (!operation.step)
        return {
            type: "startStep",
            stepKind: operation.kind === "run" ? "assistant" : "compaction",
            attempt: 1,
            contextThroughEntryId: operation.kind === "run" ? operation.inputEntryId : operation.inputThroughEntryId,
        };

    const step = operation.step;
    if (step.configurationDigest !== current.configurationDigest)
        return { type: "blocked", reason: "configuration_changed" };
    if (state.header.environmentIdentity !== current.environmentIdentity)
        return { type: "blocked", reason: "environment_changed" };
    if (step.status === "failed")
        return {
            type: "finish",
            outcome: "failed",
            error: { code: "model_error", message: "provider request failed" },
        };
    if (step.status === "attempting") {
        if (step.attempt === 2) return { type: "blocked", reason: "attempts_exhausted" };
        return {
            type: "startStep",
            stepKind: step.stepKind,
            attempt: 2,
            stepId: step.stepId,
            contextThroughEntryId: step.contextThroughEntryId,
        };
    }

    if (operation.kind === "compaction")
        return { type: "finish", outcome: "completed", finalEntryId: operation.resultEntryId };

    if (step.stopReason === "length" && assistant) {
        if (assistant.completedResults === assistant.calls.length)
            return {
                type: "finish",
                outcome: "completed",
                completion: "truncated",
                finalEntryId: step.settledEntryId,
            };
        return {
            type: "appendSynthetic",
            results: assistant.calls
                .slice(assistant.completedResults)
                .map((call, offset) =>
                    preExecutionSynthetic(
                        assistant.assistantEntryId,
                        assistant.completedResults + offset,
                        call.id,
                        call.function.name,
                        "truncated",
                    ),
                ),
        };
    }

    if (!assistant)
        return { type: "finish", outcome: "completed", completion: "normal", finalEntryId: step.settledEntryId };

    const processed = new Set(operation.toolCalls.map((tool) => `${tool.assistantEntryId}:${tool.toolIndex}`));
    for (let toolIndex = 0; toolIndex < assistant.completedResults; toolIndex++)
        processed.add(`${assistant.assistantEntryId}:${toolIndex}`);
    const untouched = assistant.calls.findIndex(
        (_, toolIndex) => !processed.has(`${assistant.assistantEntryId}:${toolIndex}`),
    );
    if (untouched >= 0) {
        const call = assistant.calls[untouched];
        const declaration = toolDeclaration(current, call.function.name);
        if (!declaration)
            return {
                type: "appendSynthetic",
                results: [
                    preExecutionSynthetic(
                        assistant.assistantEntryId,
                        untouched,
                        call.id,
                        call.function.name,
                        "unknownTool",
                    ),
                ],
            };
        let args: unknown;
        try {
            args = JSON.parse(call.function.arguments);
        } catch {
            args = undefined;
        }
        if (!args || typeof args !== "object" || Array.isArray(args))
            return {
                type: "appendSynthetic",
                results: [
                    preExecutionSynthetic(
                        assistant.assistantEntryId,
                        untouched,
                        call.id,
                        call.function.name,
                        "invalidArguments",
                    ),
                ],
            };
        return {
            type: "startTool",
            mode: "start",
            assistantEntryId: assistant.assistantEntryId,
            toolIndex: untouched,
            toolName: call.function.name,
            arguments: args as Record<string, unknown>,
        };
    }

    if (!pending.length)
        return {
            type: "startStep",
            stepKind: "assistant",
            attempt: 1,
            contextThroughEntryId: operation.toolCalls.at(-1)!.resultEntryId,
        };
    const tool = pending.sort((left, right) => left.toolIndex - right.toolIndex)[0];
    if (tool.environmentIdentity !== current.environmentIdentity)
        return { type: "blocked", reason: "environment_changed" };
    const declaration = toolDeclaration(current, tool.toolName);
    const recordedDeclaration = step.configurationSnapshot.tools.find((item) => item.name === tool.toolName);
    if (!declaration || !recordedDeclaration || declaration.definitionDigest !== recordedDeclaration.definitionDigest)
        return { type: "blocked", reason: "configuration_changed" };
    if (tool.replay === "safe" && declaration.replay === "safe" && declaration.replayKey === tool.replayKey)
        return {
            type: "startTool",
            mode: "replay",
            assistantEntryId: tool.assistantEntryId,
            toolIndex: tool.toolIndex,
            toolStartedId: tool.toolStartedId,
            toolName: tool.toolName,
            arguments: tool.arguments,
        };
    if (tool.replay === "safe" && (declaration.replay !== "safe" || declaration.replayKey !== tool.replayKey))
        return { type: "blocked", reason: "replay_declaration_changed" };
    return { type: "appendSynthetic", results: [startedSynthetic(tool, "interrupted")] };
}

File: rust/src/session_recovery.rs
use serde::Deserialize;
use serde_json::{Value, json};

use crate::session_reducer::{OperationState, SessionState, ToolCallState};

pub const SYNTHETIC_INVALID_ARGUMENTS: &str =
    "Error: Tool arguments were invalid; the tool was not executed.";
pub const SYNTHETIC_UNKNOWN_TOOL: &str = "Error: Unknown tool; the tool was not executed.";
pub const SYNTHETIC_TRUNCATED: &str = "Error: Tool call arguments were truncated by the model token limit; the tool was not executed.";
pub const SYNTHETIC_ABORTED: &str = "Operation aborted before execution.";
pub const SYNTHETIC_INTERRUPTED: &str =
    "Operation interrupted after execution status became unknown; the tool was not replayed.";

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CurrentTool {
    pub name: String,
    pub definition_digest: String,
    pub replay: String,
    pub replay_key: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CurrentConfiguration {
    pub configuration_digest: String,
    pub environment_identity: String,
    pub tools: Vec<CurrentTool>,
}

fn declaration<'a>(current: &'a CurrentConfiguration, name: &str) -> Option<&'a CurrentTool> {
    current.tools.iter().find(|tool| tool.name == name)
}

fn assistant(state: &SessionState) -> Option<(String, Vec<Value>)> {
    let step = match &state.operation {
        OperationState::Run {
            step: Some(step), ..
        } => step,
        _ => return None,
    };
    let entry_id = step.settled_entry_id.clone()?;
    let message = state.transcript.iter().rev().find(|message| {
        message.get("role").and_then(Value::as_str) == Some("assistant")
            && message
                .get("tool_calls")
                .and_then(Value::as_array)
                .is_some_and(|calls| !calls.is_empty())
    })?;
    Some((entry_id, message.get("tool_calls")?.as_array()?.clone()))
}

fn call(call: &Value) -> (&str, &str, &str) {
    (
        call.get("id").and_then(Value::as_str).unwrap_or_default(),
        call.pointer("/function/name")
            .and_then(Value::as_str)
            .unwrap_or_default(),
        call.pointer("/function/arguments")
            .and_then(Value::as_str)
            .unwrap_or_default(),
    )
}

fn synthetic(assistant_id: &str, index: usize, tool_call: &Value, reason: &str) -> Value {
    let (call_id, tool_name, _) = call(tool_call);
    json!({
        "assistantEntryId": assistant_id,
        "toolIndex": index,
        "toolCallId": call_id,
        "toolName": tool_name,
        "reason": reason,
        "content": synthetic_content(reason),
    })
}

fn interrupted(tool: &ToolCallState) -> Value {
    json!({
        "assistantEntryId": tool.assistant_entry_id,
        "toolIndex": tool.tool_index,
        "toolCallId": tool.tool_call_id,
        "toolName": tool.tool_name,
        "toolStartedId": tool.tool_started_id,
        "resultEntryId": tool.result_entry_id,
        "reason": "interrupted",
        "content": SYNTHETIC_INTERRUPTED,
    })
}

fn synthetic_content(reason: &str) -> &'static str {
    match reason {
        "invalidArguments" => SYNTHETIC_INVALID_ARGUMENTS,
        "unknownTool" => SYNTHETIC_UNKNOWN_TOOL,
        "truncated" => SYNTHETIC_TRUNCATED,
        "aborted" => SYNTHETIC_ABORTED,
        _ => SYNTHETIC_INTERRUPTED,
    }
}

pub fn plan_recovery(state: &SessionState, current: &CurrentConfiguration) -> Value {
    let operation = &state.operation;
    if matches!(operation, OperationState::Idle) {
        return json!({"type":"finish","outcome":"completed","completion":"normal"});
    }
    let assistant = assistant(state);
    let pending = match operation {
        OperationState::Run { tool_calls, .. } => tool_calls
            .iter()
            .filter(|tool| tool.status == "pending")
            .collect::<Vec<_>>(),
        _ => Vec::new(),
    };
    let (kind, step, abort_requested, context_id) = match operation {
        OperationState::Run {
            input_entry_id,
            step,
            abort_requested,
            ..
        } => ("run", step.as_ref(), *abort_requested, input_entry_id),
        OperationState::Compaction {
            input_through_entry_id,
            step,
            abort_requested,
            ..
        } => (
            "compaction",
            step.as_ref(),
            *abort_requested,
            input_through_entry_id,
        ),
        OperationState::Idle => unreachable!(),
    };
    if abort_requested {
        if step.is_some_and(|step| step.status == "attempting") {
            return json!({"type":"closeAttempt","error":{"code":"aborted","message":"Operation aborted"}});
        }
        if let (OperationState::Run { tool_calls, .. }, Some((assistant_id, calls))) =
            (operation, &assistant)
        {
            let mut results = Vec::new();
            for (index, call) in calls.iter().enumerate() {
                match tool_calls.iter().find(|tool| {
                    tool.assistant_entry_id == *assistant_id && tool.tool_index == index as u64
                }) {
                    Some(tool) if tool.status == "pending" => results.push(interrupted(tool)),
                    Some(_) => {}
                    None => results.push(synthetic(assistant_id, index, call, "aborted")),
                }
            }
            if !results.is_empty() {
                return json!({"type":"appendSynthetic","results":results});
            }
        }
        return json!({"type":"finish","outcome":"aborted"});
    }
    let Some(step) = step else {
        return json!({
            "type":"startStep",
            "stepKind": if kind == "run" { "assistant" } else { "compaction" },
            "attempt":1,
            "contextThroughEntryId":context_id,
        });
    };
    if step.status == "failed" {
        return json!({"type":"finish","outcome":"failed","error":{"code":"model_error","message":"provider request failed"}});
    }
    if step.status == "attempting" {
        if step.attempt == 2 {
            return json!({"type":"blocked","reason":"attempts_exhausted"});
        }
        if step.configuration_digest != current.configuration_digest {
            return json!({"type":"blocked","reason":"configuration_changed"});
        }
        return json!({
            "type":"startStep","stepKind":step.step_kind,"attempt":2,
            "stepId":step.step_id,"contextThroughEntryId":step.context_through_entry_id,
        });
    }
    if let OperationState::Compaction {
        result_entry_id, ..
    } = operation
    {
        return json!({"type":"finish","outcome":"completed","finalEntryId":result_entry_id});
    }
    if step.stop_reason.as_deref() == Some("length")
        && let Some((assistant_id, calls)) = &assistant
    {
        return json!({
            "type":"appendSynthetic",
            "results":calls.iter().enumerate().map(|(index, call)| synthetic(assistant_id, index, call, "truncated")).collect::<Vec<_>>()
        });
    }
    let Some((assistant_id, calls)) = &assistant else {
        return json!({"type":"finish","outcome":"completed","completion":"normal","finalEntryId":step.settled_entry_id});
    };
    let tools = match operation {
        OperationState::Run { tool_calls, .. } => tool_calls,
        _ => unreachable!(),
    };
    for (index, raw) in calls.iter().enumerate() {
        if tools
            .iter()
            .any(|tool| tool.assistant_entry_id == *assistant_id && tool.tool_index == index as u64)
        {
            continue;
        }
        let (_, name, arguments) = call(raw);
        if declaration(current, name).is_none() {
            return json!({"type":"appendSynthetic","results":[synthetic(assistant_id, index, raw, "unknownTool")]});
        }
        let Ok(Value::Object(arguments)) = serde_json::from_str::<Value>(arguments) else {
            return json!({"type":"appendSynthetic","results":[synthetic(assistant_id, index, raw, "invalidArguments")]});
        };
        return json!({
            "type":"startTool","mode":"start","assistantEntryId":assistant_id,
            "toolIndex":index,"toolName":name,"arguments":arguments,
        });
    }
    if pending.is_empty() {
        return json!({
            "type":"startStep","stepKind":"assistant","attempt":1,
            "contextThroughEntryId":tools.last().unwrap().result_entry_id,
        });
    }
    let tool = *pending.iter().min_by_key(|tool| tool.tool_index).unwrap();
    if tool.environment_identity != current.environment_identity {
        return json!({"type":"blocked","reason":"environment_changed"});
    }
    let Some(current_tool) = declaration(current, &tool.tool_name) else {
        return json!({"type":"blocked","reason":"configuration_changed"});
    };
    if current_tool.definition_digest.is_empty() {
        return json!({"type":"blocked","reason":"configuration_changed"});
    }
    if tool.replay == "safe"
        && current_tool.replay == "safe"
        && current_tool.replay_key == tool.replay_key
    {
        return json!({
            "type":"startTool","mode":"replay","assistantEntryId":tool.assistant_entry_id,
            "toolIndex":tool.tool_index,"toolStartedId":tool.tool_started_id,
            "toolName":tool.tool_name,"arguments":tool.arguments,
        });
    }
    if tool.replay == "safe" {
        return json!({"type":"blocked","reason":"replay_declaration_changed"});
    }
    json!({"type":"appendSynthetic","results":[interrupted(tool)]})
}
